### ISSUE TITLE
Confirmation buffer on shutdown

### ISSUE THREAD

**Issue by Crevil** [makasim/amqpextra#60]
Thank you for adding support for publish confirmations in #43 and #44.
I have a question regarding how the buffer is handled during shutdown.

Say there are 10 messages in-flight waiting for confirmation in the confirmation buffer and we call cancel the context passed to the dialer in `amqpextra.NewDialer(amqpextra.WithContext(ctx))`.

Will the close call block until confirmations has been received or timed out or will we abort right away?

I wasn't able to find some documentation around this and I couldn't see any tests that tests that?

**Comment by makasim** [comment]
I suppose It should send [amqp.ErrClosed](https://github.com/makasim/amqpextra/blob/350c5620b4add78833bef5721e969c72cef0a2fa/publisher/publisher.go#L373) to all pending messages, wait till it is [done](https://github.com/makasim/amqpextra/blob/350c5620b4add78833bef5721e969c72cef0a2fa/publisher/publisher.go#L320) and exit.

Ideally you should stop publishing, wait till all pending messages are published and confirmed and only after call publisher.Close()

**Comment by makasim** [comment]
I gave it another though and It looks like there is room for improvement. Here https://github.com/makasim/amqpextra/blob/master/publisher/publisher.go#L365 we can catch graceful stop and give some time for messages to catch up.

**Comment by Crevil** [comment]
Yes, that would make good sense. It would be nice with a timeout on the last catch up part, as to be able to indicate how long one is willing to wait for the issue to resolve.

**Comment by infinytum** [comment]
> Ideally you should stop publishing, wait till all pending messages are published and confirmed and only after call publisher.Close()

How would I check if there are any messages still in-flight?

**Comment by makasim** [comment]
By closing contexts in proper order. You have some inputs and outputs in your app, and a pipeline in between. So, you have to start stopping the app from inputs and go to outputs.
